File: app/src/main/java/com/thatcakeid/zrytezene/RegisterActivity.java
package com.thatcakeid.zrytezene;

import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;
import com.google.android.material.textfield.TextInputLayout;
import com.google.firebase.FirebaseApp;
import com.google.firebase.auth.FirebaseAuth;

import com.thatcakeid.zrytezene.databinding.ActivityRegisterBinding;

public class RegisterActivity extends AppCompatActivity {
    private TextInputEditText email_tie, passw_tie, passw2_tie;
    private TextInputLayout email_til, passw2_til;
    private FirebaseAuth auth;

    ActivityRegisterBinding binding;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_register);

        email_tie = binding.registerEmailTie;
        passw_tie = binding.registerPasswTie;
        passw2_tie = binding.registerPassw2Tie;

        email_til = binding.registerEmailTil;
        TextInputLayout passw_til = binding.registerPasswTil;
        passw2_til = binding.registerPassw2Til;

        MaterialButton button_continue = binding.buttonContinue;

        TextView textView5 = binding.textView5;

        FirebaseApp.initializeApp(getApplicationContext());
        auth = FirebaseAuth.getInstance();

        email_tie.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) { }

            @Override
            public void afterTextChanged(Editable s) { }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                if (email_tie.getText().toString().trim().matches("[a-zA-Z0-9._-]+@[a-z]+\\.+[a-z]+")) {
                    email_til.setError(null);
                } else {
                    email_til.setError("Invalid email!");
                }
            }
        });

        Intent intent = new Intent();
        if (intent.getStringExtra("email") != null) email_tie.setText(intent.getStringExtra("email"));

        button_continue.setOnClickListener(v -> {
            String email = email_tie.getText().toString().trim();
            String password = passw_tie.getText().toString().trim();
            String password_repeat = passw2_tie.getText().toString().trim();

            if (email.matches("[a-zA-Z0-9._-]+@[a-z]+\\.+[a-z]+")) {
                if (password.length() == 0) {
                    Toast.makeText(RegisterActivity.this, "Password can't be empty!", Toast.LENGTH_LONG).show();
                    return;
                }

                if (!password.equals(password_repeat)) {
                    passw2_til.setError("Password does not match");
                    return;
                }

                auth.createUserWithEmailAndPassword(email_tie.getText().toString().trim(),
                    passw_tie.getText().toString()).addOnSuccessListener(authResult -> {

                        // auth.getCurrentUser() will NEVER be null
                        assert auth.getCurrentUser() != null;

                        auth.getCurrentUser().sendEmailVerification().addOnSuccessListener(aVoid -> {
                            //Snackbar.make(view, "A verification link has been sent to your email. Please check your inbox or spam box.", Snackbar.LENGTH_LONG);
                            Toast.makeText(RegisterActivity.this, "A verification link has been sent to your email. Please check your inbox or spam box.", Toast.LENGTH_LONG).show();

                        }).addOnFailureListener(e -> {
                            //Snackbar.make(view, e.getMessage(), Snackbar.LENGTH_LONG);
                            Toast.makeText(RegisterActivity.this, e.getMessage(), Toast.LENGTH_LONG).show();
                        });

                        auth.signOut();

                    }).addOnFailureListener(e -> {
                        //Snackbar.make(view, e.getMessage(), Snackbar.LENGTH_LONG);
                        Toast.makeText(RegisterActivity.this, e.getMessage(), Toast.LENGTH_LONG).show();
                    });
            } else {
                //Snackbar.make(view, "Invalid email!", Snackbar.LENGTH_LONG);
                Toast.makeText(RegisterActivity.this, "Invalid email!", Toast.LENGTH_LONG).show();
            }
        });

        textView5.setOnClickListener(v -> finish());
    }
}

File: app/src/main/java/com/thatcakeid/zrytezene/RegisterActivity.kt
package com.thatcakeid.zrytezene

import com.thatcakeid.zrytezene.ExtraMetadata.setWatermarkColors
import androidx.appcompat.app.AppCompatActivity
import com.google.firebase.auth.FirebaseAuth
import android.os.Bundle
import android.text.TextWatcher
import android.text.Editable
import com.google.android.material.snackbar.Snackbar
import com.thatcakeid.zrytezene.databinding.ActivityRegisterBinding
import java.lang.Exception

class RegisterActivity : AppCompatActivity() {
    private val binding by lazy {
        ActivityRegisterBinding.inflate(layoutInflater)
    }

    private val auth by lazy { FirebaseAuth.getInstance() }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(binding.root)
        setWatermarkColors(binding.textWatermark, binding.watermarkRoot)

        binding.registerEmailTie.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
            override fun afterTextChanged(s: Editable) {}

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                if (binding.registerEmailTie.text.toString().trim()
                        .matches("[a-zA-Z0-9._-]+@[a-z]+\\.+[a-z]+".toRegex())) {
                    binding.registerEmailTie.error = null
                } else {
                    binding.registerEmailTie.error = "Invalid email!"
                }
            }

        })

        if (intent.getStringExtra("email") != null)
            binding.registerEmailTie.setText(intent.getStringExtra("email"))

        binding.buttonContinue.setOnClickListener {

            val email = binding.registerEmailTie.text.toString().trim()
            val password = binding.registerPasswTie.text.toString().trim()
            val passwordRepeat = binding.registerPassw2Tie.text.toString().trim()

            if (email.matches("[a-zA-Z0-9._-]+@[a-z]+\\.+[a-z]+".toRegex())) {
                if (password.isEmpty()) {
                    Snackbar.make(
                        binding.root,
                        "Password can't be empty!",
                        Snackbar.LENGTH_LONG
                    ).show()

                    return@setOnClickListener
                }

                if (password != passwordRepeat) {
                    Snackbar.make(
                        binding.root,
                        "Password does not match",
                        Snackbar.LENGTH_LONG
                    ).show()

                    return@setOnClickListener
                }
                auth.createUserWithEmailAndPassword(
                    binding.registerEmailTie.text.toString().trim { it <= ' ' },
                    binding.registerPasswTie.text.toString()
                ).addOnSuccessListener {
                    assert(auth.currentUser != null)
                    auth.currentUser!!
                        .sendEmailVerification().addOnSuccessListener {
                            Snackbar.make(
                                binding.root,
                                "A verification link has been sent to your email. Please check your inbox or spam box.",
                                Snackbar.LENGTH_LONG
                            ).show()
                        }
                        .addOnFailureListener { e: Exception ->
                            Snackbar.make(
                                binding.root,
                                e.message!!,
                                Snackbar.LENGTH_LONG
                            ).show()
                        }
                    auth.signOut()
                }.addOnFailureListener { e: Exception ->
                    Snackbar.make(
                        binding.root,
                        e.message!!,
                        Snackbar.LENGTH_LONG
                    ).show()
                }
            } else {
                Snackbar.make(binding.root, "Invalid email!", Snackbar.LENGTH_LONG).show()
            }
        }
        binding.textView5.setOnClickListener { finish() }
    }
}